### PR TITLE
skipping test dependent on safetensor

### PR DESCRIPTION
**Pull request by amitsrivastava78**
## Checklist

- [ ] I have added all the necessary unit tests for my change.
- [ ] I have verified that my change does not break existing code and works with all backends (TensorFlow, JAX, and PyTorch).
- [ ] My PR is based on the latest changes of the main branch (if unsure, rebase the code).
- [ ] I have followed the Keras Hub [Model contribution guidelines](https://github.com/keras-team/keras-hub/blob/master/CONTRIBUTING_MODELS.md) in making these changes.
- [ ] I have followed the Keras Hub [API design guidelines](https://github.com/keras-team/keras-hub/blob/master/API_DESIGN_GUIDE.md) in making these changes.
- [ ] I have signed the [Contributor License Agreement](https://cla.developers.google.com/about).

### CODE REVIEW

File: keras_hub/src/models/backbone_test.py [hunk 1]
```diff
@@ -107,6 +107,7 @@ def test_save_to_preset(self):
         new_out = restored_backbone(data)
         self.assertAllClose(ref_out, new_out)
 
+    @pytest.mark.skip(reason="Disabling this test for now as it needs safetensor")
```
[the remark this answers is not in the document]
Comment (amitsrivastava78) [reply keras_hub/src/models/backbone_test.py]: Done

File: keras_hub/src/models/task_test.py [hunk 2]
```diff
@@ -289,6 +289,7 @@ def _create_gemma_for_export_tests(self):
         causal_lm = GemmaCausalLM(backbone=backbone, preprocessor=preprocessor)
         return causal_lm, preprocessor
 
+    @pytest.mark.skip(reason="Disabling this test for now as it needs safetensor")
```
[the remark this answers is not in the document]
Comment (amitsrivastava78) [reply keras_hub/src/models/task_test.py]: Done

File: keras_hub/src/models/clip/clip_layers.py [hunk 3]
```diff
@@ -52,10 +52,7 @@ def build(self, input_shape):
         self.position_ids = self.add_weight(
```
Comment (james77777778) [inline keras_hub/src/models/clip/clip_layers.py:52]: I think we can use the same code block as shown below:

```python
        self.position_ids = ops.expand_dims(
            ops.arange(0, self.num_positions), axis=0
        )
```

Ref: https://github.com/huggingface/transformers/blob/main/src/transformers/models/clip/modeling_clip.py#L153

One thing to keep in mind is that we will need to update our presets after this PR is merged.

Comment (amitsrivastava78) [reply keras_hub/src/models/clip/clip_layers.py:52]: Done

File: keras_hub/src/tests/test_case.py [hunk 4]
```diff
@@ -407,7 +407,15 @@ def _get_supported_layers(mode):
...
-            revived_model.set_weights(weights)
+            revived_weights = revived_model.get_weights()
+
+            # Only attempt weight restoration if weight counts match
+            if len(weights) == len(revived_weights):
+                revived_model.set_weights(weights)
+            else:
+                # Skip weight restoration for models with dynamic structure
+                # This can happen with conditional weight creation
+                pass
```
Comment (JyotinderSingh) [inline keras_hub/src/tests/test_case.py]: Let's not modify the base quantization testing logic for a single test failure. DinoV2 (for which this is supposed to be a fix) is failing due to new behavior introduced in https://github.com/keras-team/keras/pull/21608. Simply replacing the new `re.fullmatch` call with the older and incorrect `re.search` fixes the test failure.

This points to some potential issue with the model itself, since the other models are not running into this failure. One option is to conditionally disable quantization testing for this particular model until we figure out what the issue is.

Comment (JyotinderSingh) [reply keras_hub/src/tests/test_case.py]: Screenshot 2025-09-09 at 10 33 27 AM

Did some debugging, looks like the issue is that the `revived_model` has a`dinov2_backbone_4` prefix attached to the layer paths - which ends up causing it to not match with the policies stored in the `DTypePolicyMap`.

Comment (JyotinderSingh) [reply keras_hub/src/tests/test_case.py]: One fix I can think of, is to append a wildcard matching character `.*` to the beginning of the policy keys when generating the policy map: https://github.com/keras-team/keras-hub/blob/64205cb3d37069363f6eef9b3a8df2d7b312691b/keras_hub/src/models/backbone.py#L94-L103

So instead of storing `'encoder/layers_0/attention/attention/query'` inside the policy map, we instead store `'.*encoder/layers_0/attention/attention/query'`. That should fix this issue.

@mattdangerw wdyt?

Comment (james77777778) [reply keras_hub/src/tests/test_case.py]: @JyotinderSingh 

A workaround is to explicitly set the `name`.
https://github.com/keras-team/keras-hub/blob/4b673d99cc13d817c2fba909a308ba59c4fb1389/keras_hub/src/models/dinov2/dinov2_backbone_test.py#L23

I also encountered this issue in the DepthAnything PR and fixed it the same way.

Comment (JyotinderSingh) [reply keras_hub/src/tests/test_case.py]: Thanks @james77777778! This looks like an elegant way to handle this.

On a side note, I wonder if we should still make the change to the backbone to append a `r'.*'` to the layer paths?

```py
        dtype = self.dtype_policy
        prefix = r".*"
        if not isinstance(dtype, keras.dtype_policies.DTypePolicyMap):
            policy_map = keras.dtype_policies.DTypePolicyMap()
            for layer in self._flatten_layers():
                if layer.quantization_mode is not None:
                    policy_map[prefix + layer.path] = layer.dtype_policy
            if len(policy_map) > 0:
                dtype = policy_map
        else:
            # In case we are given a DTypePolicyMap, we need to
            # append a .* to all keys to make them match
            # layer paths with different prefixes (each model
            # instance can have a new prefix).
            new_map = keras.dtype_policies.DTypePolicyMap()
            for key, value in dtype.items():
                if not key.startswith(prefix):
                    new_map[prefix + key] = value
                else:
                    new_map[key] = value
            dtype = new_map
```

Comment (james77777778) [reply keras_hub/src/tests/test_case.py]: This seems a bit complicated.

I'm wondering if it would be a good idea to add a new attr such as `Variable._policy_path` and include some logic in `auto_name` to get a regex-compatible path in the core Keras.

For KerasHub, we could then migrate to use `._policy_path` if it exists and fall back to `.path` otherwise. For now, we can explicitly set `name` to pass the tests.

@JyotinderSingh @mattdangerw  WDYT?

Comment (amitsrivastava78) [reply keras_hub/src/tests/test_case.py]: @james77777778  @JyotinderSingh  thanks for the comments, have incorporated it now!
@mattdangerw let me know if you have any other suggestion for this

Comment (amitsrivastava78) [reply keras_hub/src/tests/test_case.py]: Had to update the keras_hub/src/tests/test_case.py file as this was still failing

File: keras_hub/src/models/backbone_test.py [hunk 5]
```diff
@@ -3,6 +3,11 @@
 import numpy as np
 import pytest
 
+try:
+    import safetensors
```
Comment (mattdangerw) [inline keras_hub/src/models/backbone_test.py]: we should not need the conditional import for tests, just for library code. (if the worry is google side we have other solutions we can pursue there)

File: keras_hub/src/models/clip/clip_layers.py [hunk 6]
```diff
@@ -49,18 +49,11 @@ def build(self, input_shape):
             dtype=self.variable_dtype,
             name="class_embedding",
         )
-        self.position_ids = self.add_weight(
```
Comment (mattdangerw) [inline keras_hub/src/models/clip/clip_layers.py:52]: what's the concern with this code? this will break checkpoint compatibility, so we might want to avoid unless we are planning to regen all the clip checkpoints

File: keras_hub/src/models/siglip/siglip_layers.py [hunk 7]
```diff
@@ -67,18 +67,8 @@ def __init__(
         )
 
     def build(self, input_shape):
-        self.position_ids = self.add_weight(
```
Comment (mattdangerw) [inline keras_hub/src/models/siglip/siglip_layers.py]: same here, let's avoid this if we can. changing would break checkpoint compat

File: keras_hub/src/models/task_test.py [hunk 8]
```diff
@@ -6,6 +6,11 @@
 import pytest
 from absl.testing import parameterized
 
+try:
+    import safetensors
```
Comment (mattdangerw) [inline keras_hub/src/models/task_test.py]: no need to conditional import for tests

File: keras_hub/src/tests/test_case.py (hunk 4 above)
Comment (JyotinderSingh) [reply keras_hub/src/tests/test_case.py]: I tried the test locally, and `keras_hub/src/models/dinov2/dinov2_backbone_test.py::DINOV2BackboneTest::test_backbone_basics` seems to be passing even without the changes in `keras_hub/src/tests/test_case.py`. Looks like the changes in `keras_hub/src/models/dinov2/dinov2_backbone_test.py` are enough.

Unsure why this failed on the CI.

Comment (mattdangerw) [reply keras_hub/src/tests/test_case.py]: Let's not submit what's here currently, as it's subverting the intent of the test. Cloning a model then calling get_weights/set_weights should preserve the weight count. Probably better to just disable individual tests or parts of tests that are failing for now. We can always `run_quantization_check=False` to a backbone test with a todo while we fix things.

@james77777778 @JyotinderSingh re suggestions, is the issue just a bug in how names are propagated? Cloning a model via `from_config(get_config())` or saving (which will do `from_config(get_config())` among other things), should not update layer names. I wonder if we are just facing a bug in name serialization/deserialization. But I may be missing something. I do see how there's a separate question of the autoname functionality and usability with that when specifying paths. But for this test and our creation of a dtype policy map during serialzation I think names should always be preserved? Let me know if I am missing something! Very possible.

Comment (james77777778) [reply keras_hub/src/tests/test_case.py]: If `name` isn't specified, the model (especially `Backbone`) will be auto-named, which can lead to inconsistent naming issue during serialization and deserialization.

Here is a demo colab:
https://colab.research.google.com/drive/1XS3Er0CFXszDQf72-Zr8aAyRChNp1ju-?usp=sharing
image

We may need to explicitly set `name` for our presets and for our tests to avoid this.

Another question is whether we should support name-insensitive `DTypePolicyMap`. Currently, the `DTypePolicyMap` is likely to fail when it encounters auto-named layers.

Comment (mattdangerw) [reply keras_hub/src/tests/test_case.py]: > Another question is whether we should support name-insensitive DTypePolicyMap. Currently, the DTypePolicyMap is likely to fail when it encounters auto-named layers.

Agreed this is something we should consider, but somewhat orthogonal to the test fix question.

> If name isn't specified, the model (especially Backbone) will be auto-named, which can lead to inconsistent naming issue during serialization and deserialization.

I don't follow this though. If the name isn't supplied, the autoname will still be in the config right? So the deserialized version will always have the same (potentially autonamed) name as the original. https://colab.sandbox.google.com/gist/mattdangerw/5932aef63edd6abb83c508cf34914f49/autonaming.ipynb Do our tests require a fixed name for all test runs? Shouldn't they just need a consistent name between serialized and deserialized models?

Comment (james77777778) [reply keras_hub/src/tests/test_case.py]: > Agreed this is something we should consider, but somewhat orthogonal to the test fix question.

I can propose a PR for this.

> I don't follow this though. If the name isn't supplied, the autoname will still be in the config right? So the deserialized version will always have the same (potentially autonamed) name as the original.

You are right. During `from_config(get_config())`, the `name` will be propagated, which means the only issue is that we need to provide `name` in our test suite.

It is still a good practice to provide `name` in our conversion scripts to avoid unexpected naming issues. For example, if a conversion script instantiates two backbones, the uploaded model's `name` might contain an unexpected `_1` suffix.

File: keras_hub/src/models/backbone_test.py (hunk 5 above)
Comment (amitsrivastava78) [reply keras_hub/src/models/backbone_test.py]: Sure

File: keras_hub/src/models/task_test.py (hunk 8 above)
Comment (amitsrivastava78) [reply keras_hub/src/models/task_test.py]: Sure

File: keras_hub/src/models/siglip/siglip_layers.py (hunk 7 above)
Comment (amitsrivastava78) [reply keras_hub/src/models/siglip/siglip_layers.py]: Sure

File: keras_hub/src/tests/test_case.py (hunk 4 above)
Comment (amitsrivastava78) [reply keras_hub/src/tests/test_case.py]: ok have run_quantization_check=False for now, will wait for the PR from @JyotinderSingh and @james77777778 to fix this

File: keras_hub/src/models/deberta_v3/disentangled_self_attention.py [hunk 9]
```diff
@@ -237,13 +237,13 @@ def _get_log_pos(abs_pos, mid):
...
         )
-        bucket_pos = ops.cast(bucket_pos, dtype="int")
+        bucket_pos = ops.cast(bucket_pos, dtype="int32")
 
         return bucket_pos
 
     def _get_rel_pos(self, num_positions):
         ids = ops.arange(num_positions)
-        ids = ops.cast(ids, dtype="int")
+        ids = ops.cast(ids, dtype="int32")
```
Comment (JyotinderSingh) [inline keras_hub/src/models/deberta_v3/disentangled_self_attention.py]: I'm concerned that converting `"int"` to `"int32"` might not be the ideal solution.

In TensorFlow, `"int32"` tensors are typically placed on the CPU, which can cause issues with XLA graph generation as parts of the model would reside on different devices. `"int"` usually maps to `"int64"` and stays on the GPU.

Could you help me understand the original problem we were facing here? Maybe we can find a more robust fix. 

Also, we need to confirm if this change is passing on the GPU CI.

reference: https://stackoverflow.com/questions/37439299/no-gpu-kernel-for-an-int32-variable-op/37452938#37452938

Comment (amitsrivastava78) [reply keras_hub/src/models/deberta_v3/disentangled_self_attention.py]: Yes they are passing on GPU, ran the test with label kokoro:force-run

Comment (hertschuh) [reply keras_hub/src/models/deberta_v3/disentangled_self_attention.py]: But maybe this forces the computation to be executed on CPU, the test would still pass.

Comment (amitsrivastava78) [reply keras_hub/src/models/deberta_v3/disentangled_self_attention.py]: Yes , i think you guys are right, removed the changes

File: keras_hub/src/models/dinov2/dinov2_backbone_test.py [hunk 10]
```diff
@@ -35,7 +36,7 @@ def test_backbone_basics(self):
             init_kwargs=self.init_kwargs,
             input_data=self.input_data,
             expected_output_shape=(2, sequence_length, hidden_dim),
-            run_quantization_check=False,
+            run_quantization_check=False,  # TODO: Fix weight count mismatch
```
Comment (JyotinderSingh) [inline keras_hub/src/models/dinov2/dinov2_backbone_test.py:39]: We can remove this change, since it was already addressed in #2397 for a Gemma release.

File: keras_hub/src/models/dinov2/dinov2_backbone_test.py [hunk 11]
```diff
@@ -127,7 +128,7 @@ def test_backbone_basics(self):
...
```
Comment (JyotinderSingh) [inline keras_hub/src/models/dinov2/dinov2_backbone_test.py:131]: We can remove this change, since it was already addressed in #2397 for a Gemma release.

File: keras_hub/src/tests/test_case.py [hunk 12]
```diff
@@ -381,7 +381,6 @@ def _get_supported_layers(mode):
                     )
             # Ensure the correct `dtype` is set for sublayers or submodels in
             # `init_kwargs`.
-            original_init_kwargs = init_kwargs.copy()
```
Comment (JyotinderSingh) [inline keras_hub/src/tests/test_case.py:384]: Why did we remove this?

File: keras_hub/src/tests/test_case.py [hunk 13]
```diff
@@ -408,8 +407,6 @@ def _get_supported_layers(mode):
             # Check weights loading.
             weights = model.get_weights()
             revived_model.set_weights(weights)
-            # Restore `init_kwargs`.
-            init_kwargs = original_init_kwargs
```
Comment (JyotinderSingh) [inline keras_hub/src/tests/test_case.py:412]: Why did we remove this?

**Comment by james77777778** [comment]
@JyotinderSingh @amitsrivastava78 

I don't think this PR is effective as the issue was already resolved by #2397.

One quick question: is `dtype=int` causing any issues? I saw this PR was making changes like this. If so, I can submit a PR to address it but we might need to re-upload the presets to reflect the changes.

File: keras_hub/src/tests/test_case.py (hunk 12 above)
Comment (mattdangerw) [reply keras_hub/src/tests/test_case.py:384]: Looking at the code it looks like this was supposed to prevent mutating an original dict in this function. But it won't work on master or in this PR. We shouldn't copy a dict and try to assign it back at the end. That assignment won't affect the calling dict. We should instead make a copy and work from it in this function. I'll just update as I merge this.

Comment (mattdangerw) [reply keras_hub/src/tests/test_case.py:384]: ah just saw @james77777778's comment, sounds like we no longer need these changes.
